src/ir/auth_logic/ast.h

```diff
@@ -139,6 +139,55 @@ class Assertion {
   AssertionVariantType value_;
 };
 
+enum class ConstantType { kNumberType, kPrincipalType };
+
+// Relation declaration types are of 3 forms
+// 1. NumberType
+// 2. PrincipalType
+// 3. User defined CustomType (string to store name of the type)
+class ArgumentType {
+ public:
+  enum class Kind { kNumber, kPrincipal, kCustom };
+  explicit ArgumentType(Kind kind, std::string name)
+      : kind_(std::move(kind)), name_(std::move(name)) {}
+  const Kind& kind() const { return kind_; }
+  const std::string name() const { return name_; }
```

Comment:
`const std::string&` here. Otherwise, you will make an unnecessary copy. You can also use `absl::string_view` above.